File: src/health.ts
/**
 * Approval signal health — detects patterns that indicate human oversight
 * has degraded into rubber-stamping. Guards the guards.
 *
 * When health degrades, the pipeline should pause trust graduation and
 * escalate to a supervisor rather than treating approvals as valid signal.
 */

export type HealthFlag =
  /** More than 3 decisions recorded in under 30 seconds. */
  | 'rapid_fire'
  /** More than 5 approvals recorded in under 60 seconds. */
  | 'batch_approval'
  /** Every decision in the last 20+ was an approval — zero variance. */
  | 'no_variance'
  /** More than 30% of the last 10 decisions were rejections (sudden reversal). */
  | 'dismiss_spike'

export interface HealthReport {
  /** true when no flags are raised. */
  healthy: boolean
  /** Active flags explaining why health is degraded. */
  flags: HealthFlag[]
  /** Number of decisions in the current analysis window. */
  window_size: number
  /** Approval rate across the full window (0–1). */
  approval_rate: number
  assessed_at: string
}

interface Decision {
  timestamp: number
  approved: boolean
}

export interface ApprovalSignalHealthOptions {
  /**
   * Maximum decisions kept in memory. Older entries are pruned.
   * Default 200.
   */
  max_window: number
  now?: () => number
}

/**
 * Monitors the stream of approval decisions and surfaces patterns that
 * indicate oversight has degraded.
 *
 * Healthy oversight has variance (some rejections), is paced (not instant),
 * and is engaged (not rubber-stamping everything).
 *
 * @example
 * ```ts
 * const health = new ApprovalSignalHealth()
 *
 * // After each human decision:
 * health.record(true)   // approved
 * health.record(false)  // rejected
 *
 * const report = health.assess()
 * if (!report.healthy) {
 *   console.warn('Oversight degraded:', report.flags)
 *   // pause trust graduation, alert supervisor
 * }
 * ```
 */
export class ApprovalSignalHealth {
  private window: Decision[] = []
  private opts: Required<ApprovalSignalHealthOptions>

  constructor(options: Partial<ApprovalSignalHealthOptions> = {}) {
    this.opts = {
      max_window: options.max_window ?? 200,
      now: options.now ?? (() => Date.now()),
    }
  }

  /** Record a human approval or rejection. Call after every gate decision. */
  record(approved: boolean): void {
    this.window.push({ timestamp: this.opts.now(), approved })
    if (this.window.length > this.opts.max_window) {
      this.window.shift()
    }
  }

  /** Assess current oversight health and return a report. */
  assess(): HealthReport {
    const now = this.opts.now()
    const flags: HealthFlag[] = []

    const recent30s = this.window.filter((d) => now - d.timestamp < 30_000)
    const recent60s = this.window.filter((d) => now - d.timestamp < 60_000)
    const last20 = this.window.slice(-20)
    const last10 = this.window.slice(-10)

    if (recent30s.length > 3) flags.push('rapid_fire')

    const approvals60s = recent60s.filter((d) => d.approved).length
    if (approvals60s > 5) flags.push('batch_approval')

    if (last20.length >= 20 && last20.every((d) => d.approved)) flags.push('no_variance')

    const rejections10 = last10.filter((d) => !d.approved).length
    if (last10.length >= 10 && rejections10 / last10.length > 0.3) flags.push('dismiss_spike')

    const approvals = this.window.filter((d) => d.approved).length
    const approval_rate = this.window.length > 0 ? approvals / this.window.length : 1

    return {
      healthy: flags.length === 0,
      flags,
      window_size: this.window.length,
      approval_rate,
      assessed_at: new Date(now).toISOString(),
    }
  }

  /** Snapshot of the raw decision window, newest last. */
  snapshot(): ReadonlyArray<Readonly<Decision>> {
    return this.window
  }
}

File: python/suretyai/health.py
from __future__ import annotations

import time
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Callable, Literal, Optional

HealthFlag = Literal["rapid_fire", "batch_approval", "no_variance", "dismiss_spike"]


@dataclass
class HealthReport:
    healthy: bool
    flags: list[HealthFlag]
    window_size: int
    approval_rate: float
    assessed_at: str


@dataclass
class _Decision:
    timestamp: float
    approved: bool


class ApprovalSignalHealth:
    """
    Monitors the stream of human approval decisions and surfaces patterns
    that indicate oversight has degraded into rubber-stamping.

    Example::

        health = ApprovalSignalHealth()

        health.record(True)
        health.record(False)

        report = health.assess()
        if not report.healthy:
            print("Oversight degraded:", report.flags)
    """

    def __init__(
        self,
        max_window: int = 200,
        now: Optional[Callable[[], float]] = None,
    ) -> None:
        self._window: list[_Decision] = []
        self._max_window = max_window
        self._now = now or time.time

    def record(self, approved: bool) -> None:
        self._window.append(_Decision(timestamp=self._now(), approved=approved))
        if len(self._window) > self._max_window:
            self._window.pop(0)

    def assess(self) -> HealthReport:
        now = self._now()
        flags: list[HealthFlag] = []

        recent_30s = [d for d in self._window if now - d.timestamp < 30]
        recent_60s = [d for d in self._window if now - d.timestamp < 60]
        last20 = self._window[-20:]
        last10 = self._window[-10:]

        if len(recent_30s) > 3:
            flags.append("rapid_fire")

        approvals_60s = sum(1 for d in recent_60s if d.approved)
        if approvals_60s > 5:
            flags.append("batch_approval")

        if len(last20) >= 20 and all(d.approved for d in last20):
            flags.append("no_variance")

        rejections_10 = sum(1 for d in last10 if not d.approved)
        if len(last10) >= 10 and rejections_10 / len(last10) > 0.3:
            flags.append("dismiss_spike")

        approvals = sum(1 for d in self._window if d.approved)
        approval_rate = approvals / len(self._window) if self._window else 1.0

        return HealthReport(
            healthy=len(flags) == 0,
            flags=flags,
            window_size=len(self._window),
            approval_rate=approval_rate,
            assessed_at=datetime.fromtimestamp(now, tz=timezone.utc).isoformat(),
        )

    def snapshot(self) -> list[_Decision]:
        return list(self._window)
